tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/RuntimeBrokerTests.cs

```diff
@@ -265,6 +264,70 @@ public async Task WamAddDefaultScopesWhenNoScopesArePassedAsync(string scopes)
 
             Assert.IsTrue(!string.IsNullOrEmpty(ex.ErrorCode));
         }
+
+        [RunOn(TargetFrameworks.NetStandard | TargetFrameworks.NetCore)]
+        public async Task WamUsernamePasswordPopTokenEnforcedWithCaOnValidResourceAsync()
+        {
+            //Arrange
+            var labResponse = await LabUserHelper.GetDefaultUserAsync().ConfigureAwait(false);
+            
+            string popUser = "popUser@msidlab4.onmicrosoft.com";
+
+            string[] scopes = { "https://msidlab4.sharepoint.com/user.read" };
+
+            IntPtr intPtr = GetForegroundWindow();
+
+            Func<IntPtr> windowHandleProvider = () => intPtr;
+
+            IPublicClientApplication pca = PublicClientApplicationBuilder
+               .Create(labResponse.App.AppId)
+               .WithParentActivityOrWindow(windowHandleProvider)
+               .WithAuthority(labResponse.Lab.Authority, "organizations")
+               .WithBroker(new BrokerOptions(BrokerOptions.OperatingSystems.Windows))
+               .Build();
+
+            // Acquire token using username password with POP on a valid resource
+            // CA policy enforces token issuance to popUser only for SPO
+            // https://learn.microsoft.com/en-us/azure/active-directory/conditional-access/concept-token-protection
+            var result = await pca.AcquireTokenByUsernamePassword(scopes, popUser, labResponse.User.GetOrFetchPassword())
+                .WithProofOfPossession("some_nonce", System.Net.Http.HttpMethod.Get, new Uri(pca.Authority))
+                .ExecuteAsync()
+                .ConfigureAwait(false);
+
+            //Act
+            Assert.AreEqual(popUser, result.Account.Username);
+        }
+
+        [RunOn(TargetFrameworks.NetStandard | TargetFrameworks.NetCore)]
+        [ExpectedException(typeof(MsalUiRequiredException))]
+        public async Task WamUsernamePasswordPopTokenEnforcedWithCaOnInValidResourceAsync()
+        {
+            //Arrange
+            var labResponse = await LabUserHelper.GetDefaultUserAsync().ConfigureAwait(false);
+
+            string popUser = "popUser@msidlab4.onmicrosoft.com";
+
+            string[] scopes = { "user.read" };
+
+            IntPtr intPtr = GetForegroundWindow();
+
+            Func<IntPtr> windowHandleProvider = () => intPtr;
+
+            IPublicClientApplication pca = PublicClientApplicationBuilder
+               .Create(labResponse.App.AppId)
+               .WithParentActivityOrWindow(windowHandleProvider)
+               .WithAuthority(labResponse.Lab.Authority, "organizations")
+               .WithBroker(new BrokerOptions(BrokerOptions.OperatingSystems.Windows))
+               .Build();
+
+            // Acquire token using username password with POP on a resource not in the CA policy
+            // CA policy enforces token issuance to popUser only for SPO this call will fail with UI Required Exception
```

Comment:
Why would attempting a POP token request on a non-POP resource result in a UI Required error? Is that the way AAD tells the app developer "hey you do not need to use POP for this resource"? But even if that is the case, why our negative test case here would need to test that behavior?

I thought the negative test case should be to attempt a bearer token request for the same POP-protected resource, and expect an error, which in turn proves that our happy-path test case did the POP right.

Reply:
Great question. My hypothesis is that you attempt non-interactive flows like ROPC or refresh_token, then AAD will ask you to switch to interactive just to show the end-user a friendly message like "You can't get from here to here" or "This version of the app doesn't meet the security requirements". An end user who uses VS can for example switch between browser and WAM and "fix" the CA policy themselves?

Reply:
Agreed on the nagative test, that's a nice integration scenario.

Reply:
> Great question. My hypothesis is that you attempt non-interactive flows like ROPC or refresh_token, then AAD will ask you to switch to interactive just to show the end-user a friendly message like "You can't get from here to here" or "This version of the app doesn't meet the security requirements". An end user who uses VS can for example switch between browser and WAM and "fix" the CA policy themselves?

As of this writing, the test case here is *sending POP request to a non-POP resource* (MS Graph). Not sure whether ESTS would emit an error "this version of the app EXCEEDS the security requirements (i.e. you are too secure, so I'll block you)". Regardless, this is perhaps not the scenario we need to test here.

Reply:
No, ESTS is not aware of RP capability to accept various type of tokens. RP is supposed to issue `WWW-Authenticate Bearer` and `WWW-Authenticate POP` for inform the client of what auth schemes it supports.